polling_stations/apps/data_importers/data_quality_report.py

```diff
@@ -0,0 +1,523 @@
+            else:
+                self.counts[">1"] = self.counts[">1"] + 1
+
+    def get_districts_containing_zero_stations(self):
+        return self.counts["0"]
+
+    def get_districts_containing_one_stations(self):
+        return self.counts["1"]
+
+    def get_districts_containing_more_stations(self):
+        return self.counts[">1"]
+
+
+# data quality stats for UPRNs assigned polling station ids
+class AddressReport:
+    def __init__(self, council_id):
+        self.council_id = council_id
+
+    def get_uprns_in_addressbase(self):
+        return UprnToCouncil.objects.filter(lad=self.council_id).count()
+
+    def get_addresses_with_station_id(self):
+        return (
+            UprnToCouncil.objects.filter(
+                lad=self.council_id, polling_station_id__isnull=False
+            )
+            .exclude(polling_station_id="")
+            .count()
+        )
+
+    def get_addresses_without_station_id(self):
+        return UprnToCouncil.objects.filter(
+            Q(polling_station_id__isnull=True) | Q(polling_station_id=""),
+            council_id=self.council_id,
+        ).count()
+
+    def get_addresses_with_valid_station_id_ref(self):
+        cursor = connection.cursor()
+        cursor.execute(
+            """
+            SELECT COUNT(*) FROM addressbase_uprntocouncil
+            WHERE polling_station_id IN
+                (SELECT internal_council_id FROM pollingstations_pollingstation
+                WHERE council_id=%s)
+            AND lad=%s
+            AND polling_station_id != ''
+            AND polling_station_id IS NOT NULL;
+            """,
+            [self.council_id, self.council_id],
+        )
+        results = cursor.fetchall()
+        return results[0][0]
+
+    def get_addresses_with_invalid_station_id_ref(self):
+        cursor = connection.cursor()
+        cursor.execute(
+            """
+            SELECT COUNT(*) FROM addressbase_uprntocouncil
+            WHERE polling_station_id NOT IN
+                (SELECT internal_council_id FROM pollingstations_pollingstation
+                WHERE council_id=%s)
+            AND lad=%s
+            AND polling_station_id != ''
+            AND polling_station_id IS NOT NULL;
+            """,
+            [self.council_id, self.council_id],
+        )
+        results = cursor.fetchall()
+        return results[0][0]
+
+
+# generate all the stats
+class DataQualityReportBuilder:
+    def __init__(self, council_id, expecting_districts):
+        self.council_id = council_id
+        self.report = []
+        # Whether the importer is expected to have imported districts;
+        # controls whether relevant summaries appear in the report.
+        self.expecting_districts = expecting_districts
+
+    def build_header(self):
+        self.report.append("==================================")
+        self.report.append(ANSI.bold("        DATA QUALITY REPORT"))
+        self.report.append("==================================\n")
```

Comment:
Definitely not one for this PR, but just seeing this reminds me I recently had a go with https://github.com/willmcgugan/rich and its _amazing_ for rich console interfaces (colours, tables, etc). We could make some really big improvements to this module with that.

Reply:
https://github.com/DemocracyClub/UK-Polling-Stations/issues/2917